### PR TITLE
Enable using information about whether functions return constants

### PR DESCRIPTION
**Pull request by emjin**
This will help DeepSemgrep

PR checklist:

- [ ] Documentation is up-to-date
- [ ] Changelog is up-to-date
- [ ] Change has no security implications (otherwise, ping security team)

### CODE REVIEW

File: semgrep-core/src/analyzing/Dataflow_svalue.mli [hunk 1]
```diff
@@ -1,5 +1,7 @@
 type mapping = AST_generic.svalue Dataflow_core.mapping
+type constness_type = Constant | NotAlwaysConstant [@@deriving show]
```
Comment (emjin) [inline semgrep-core/src/analyzing/Dataflow_svalue.mli:6]: Could reuse G.Cst? But might be nice to separate this out

File: semgrep-core/src/analyzing/Constant_propagation.mli [hunk 2]
```diff
@@ -12,4 +12,8 @@ val constant_propagation_and_evaluate_literal :
  *)
 (* !Note that this assumes Naming_AST.resolve has been called before! *)
 val propagate_basic : Lang.t -> AST_generic.program -> unit
+
+val propagate_dataflow_one_function :
```
Comment (aryx) [inline semgrep-core/src/analyzing/Constant_propagation.mli:19]: Could be nice to document the parameter, especially what are those IL.name list?

File: semgrep-core/src/analyzing/Dataflow_svalue.mli (hunk 1 above)
Comment (aryx) [reply semgrep-core/src/analyzing/Dataflow_svalue.mli:6]: then maybe good to add a comment saying you know about G.Cst, but that you deliberately defined another one.

**Review by aryx — APPROVED** [review]

File: semgrep-core/src/analyzing/Dataflow_svalue.ml [hunk 3]
```diff
@@ -51,10 +53,77 @@ let warning tok s =
...
 
+let hook_constness_table_of_functions = ref None
+
+let result_of_function_call_is_constant lang f args =
+  let check_f f_name =
+    match !hook_constness_table_of_functions with
+    | Some constness_f -> (
+        match constness_f f_name with
+        | Some Constant ->
+            logger#info "%s is always constant" f_name;
```
Comment (IagoAbal) [inline semgrep-core/src/analyzing/Dataflow_svalue.ml]: Better use `logger#trace` or else the output of `semgrep --debug` becomes very large, and most of the times this info won't be of much use. Same for the other two uses of `logger#info` below.

File: semgrep-core/src/analyzing/Dataflow_svalue.ml [hunk 4]
```diff
@@ -51,10 +53,77 @@ let warning tok s =
...
+            true
+        | Some NotAlwaysConstant ->
+            logger#info "%s is not always constant" f_name;
+            false
+        | None ->
+            logger#info "%s has not been read" f_name;
```
Comment (IagoAbal) [inline semgrep-core/src/analyzing/Dataflow_svalue.ml]: Nitpicking here but what does `"not been read"` mean in this context? That we have not yet processed it? I was expecting a message more like `"we have no information about the constness of %s"`.

**Review by IagoAbal — APPROVED** [review]
LGTM!

File: semgrep-core/src/analyzing/Dataflow_svalue.ml (hunk 4 above)
Comment (emjin) [reply semgrep-core/src/analyzing/Dataflow_svalue.ml]: I was thinking of it as "we haven't read through the function body", but your wording makes sense

**Pull request merged**